src/JuMPContainer.jl

```diff
@@ -183,13 +185,86 @@ Base.length(it::ValueIterator)  = length(it.x)
 
 type KeyIterator{JA<:JuMPArray}
     x::JA
+    dim::Int
+    next_k_cache::Array{Any,1}
+    function KeyIterator(d)
+        n = ndims(d.innerArray)
+        new(d, n, Array(Any, n+1))
+    end
+end
+
+KeyIterator{JA}(d::JA) = KeyIterator{JA}(d)
+
+immutable IndexableIndexsets end
+immutable NotIndexableIndexsets end
```

Comment:
Personally I would prefer booleans, as this code is already a bit difficult for me to parse. We can always add complexity later on as-needed